#######################################################################################
###                                                                                 ###
###    Copyright (C) 2020,  Burair Alsaihati                                        ###
###                                                                                 ###
###    This program is free software: you can redistribute it and/or modify         ###
###    it under the terms of the GNU General Public License as published by         ###
###    the Free Software Foundation version 3                                       ###
###                                                                                 ###
###    This program is distributed in the hope that it will be useful,              ###
###    but WITHOUT ANY WARRANTY; without even the implied warranty of               ###
###    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                ###
###    GNU General Public License for more details.                                 ###
###                                                                                 ###
###    You should have received a copy of the GNU General Public License            ###
###    along with this program.  If not, see <https://www.gnu.org/licenses/>.       ###
###                                                                                 ###
###    Email: burair.alsaihati25@uga.edu, burair_99@yahoo.com, szhao@uga.edu        ###
###                                                                                 ###
#######################################################################################

# Run this shell script for each normal and tumor WES sample to generate a depth of coverage file, a germline variants file (VCF) and an annotated version of the germline variants
# Required software: GATK 3, ANNOVAR (ANNOtate VARiation) Perl package and an in-house-developed java package: Pancancer.jar
# GATK tools used: DepthOfCoverage, HaplotypeCaller and VariantFiltration
# GNU/Linux used: awk
# ANNOVAR tools used: convert2annovar.pl and annotate_variation.pl
# Pancancer.jar tools used: util.AddGeneNameCustomized

######## expected input arguments ########
# path to indexed reference genome
ref_genome=canFam3.fa
# path to GTF format reference genome annotation
gtf_file=Canis_familiaris.CanFam3.1.99.chr.gtf
# an interval list file for coding sequence (CDS) regions (generated by make_CDS_interval_list.sh)
CDS_interval_list_file=Canis_familiaris.CanFam3.1.99.gtf-chr1-38X-CDS.interval_list
# path to ANNOVAR directory containing convert2annovar.pl and annotate_variation.pl
annovar_path=annovar_CanFam3.1.99.gtf
# an id corresponding to a tumor or a normal sample
sample_id=SAMN04002427
# path to sorted, indexed, duplication marked and realigned bam file for the sample
bam_file=${sample_id}_rg_added_sorted_dedupped_removed.realigned.bam


####### GATK Depth of Coverage ######
time java -jar GenomeAnalysisTK.jar -T DepthOfCoverage \
 -R $ref_genome \
 -I $bam_file \
 --minBaseQuality 10 \
 --minMappingQuality 10 \
 -L $CDS_interval_list_file \
 -o ${sample_id}_DepthofCoverage_CDS.bed 

####### GATK germline variant calling and filtration ######
# Variant calling
time java -jar GenomeAnalysisTK.jar -T HaplotypeCaller \
 -R $ref_genome \
 -I $bam_file \
 -dontUseSoftClippedBases -stand_call_conf 20.0 \
 -o ${bam_file}.vcf

# Variant filtering
time java -jar GenomeAnalysisTK.jar -T VariantFiltration \
 -R $ref_genome \
 -V ${bam_file}.vcf \
 -filterName FS -filter "FS > 30.0" -filterName QD -filter "QD < 2.0" \
 -o ${bam_file}.filtered.vcf


################ Annovar ###################
# Extract PASS records from filtered vcf
awk '$7 == "PASS" {print $0}' ${bam_file}.filtered.vcf > ${bam_file}.filtered.vcf-PASS
# annovar input preparation
perl $annovar_path/convert2annovar.pl -format vcf4old ${bam_file}.filtered.vcf-PASS > ${bam_file}.filtered.vcf-PASS-avinput
# annovar annotate
perl $annovar_path/annotate_variation.pl --buildver canFam3 ${bam_file}.filtered.vcf-PASS-avinput $annovar_path


############## Pancancer.jar add gene names ################
# add gene names to annovar output
java -cp Pancancer.jar util.AddGeneNameCustomized ${bam_file}.filtered.vcf-PASS-avinput.exonic_variant_function $gtf_file ${bam_file}.filtered.vcf-PASS-avinput.exonic_variant_function_WithGeneName 3 3
